aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -489,10 +516,90 @@ Tensor dense_to_sparse_csc(const Tensor& self) {
+  //
+  // Given matrix
+  //
+  //  1  2  3  4
+  //  5  6  7  8
+  //  9 10 11 12
+  // 14 15 16 17
+  //
+  // _tile_tensor(matrix, {2, 2}) will yield the following 2 by 2 blocks
+  //
+  //  1  2 |  3  4 |  9 10 | 11 12
+  //  5  6 |  7  8 | 14 15 | 16 17
+  //
+  //  via a 3D Tensor of shape (4, 2, 2)
+  //
+  auto block_size_0 = self.size(0) == 0 ? 0 : self.size(0) / blocksize[0];
+  auto block_size_1 = self.size(1) == 0 ? 0 : self.size(1) / blocksize[1];
+  return self.reshape({block_size_0, blocksize[0], block_size_1, blocksize[1]})
+      .transpose(1, 2)
+      .contiguous();
+}
+
+std::pair<Tensor, Tensor> _not_zero_mask_to_col_row_indices(
+    Tensor not_zero_mask) {
+  auto col_indices = at::native::arange(not_zero_mask.size(-1))
+                         .view({1, not_zero_mask.size(-1)})
+                         .expand_as(not_zero_mask)
+                         .masked_select(not_zero_mask);
+  auto row_indices = at::native::arange(not_zero_mask.size(-2))
+                         .view({not_zero_mask.size(-2), 1})
+                         .expand_as(not_zero_mask)
+                         .masked_select(not_zero_mask);
+  return std::pair<Tensor, Tensor>(col_indices, row_indices);
+}
+
 Tensor dense_to_sparse_bsr(const Tensor& self, IntArrayRef blocksize) {
-  AT_ERROR(
-      "Conversion from ", self.layout(), " to SparseBsr is currently not supported.");
-  return self;
+  TORCH_CHECK(self.dim() == 2, "Can only covert 2D Tensor to BSR.");
+  TORCH_CHECK(
+      blocksize[0] > 0 && blocksize[1] > 0,
+      "blocksize needs to be non zero, but got ",
+      blocksize);
+  TORCH_CHECK(
+      self.size(0) % blocksize[0] == 0,
+      "Tensor size(0) ",
+      self.size(0),
+      " needs to be divisible by blocksize[0] ",
+      blocksize[0]);
+  TORCH_CHECK(
+      self.size(1) % blocksize[1] == 0,
+      "Tensor size(1) ",
+      self.size(1),
+      " needs to be divisible by blocksize[1] ",
+      blocksize[1]);
+  auto block_size_0 = self.size(0) / blocksize[0];
+
+  auto values = _tile_tensor(self, blocksize);
+  auto not_zero_mask = _tile_tensor((self != 0), blocksize);
+  // Find tiles that have at least 1 non-zero value in them.
+  not_zero_mask = not_zero_mask.any(-1).any(-1);
+  Tensor col_indices;
+  Tensor row_indices;
+  std::tie(col_indices, row_indices) =
+      _not_zero_mask_to_col_row_indices(not_zero_mask);
+  Tensor crow_indices = at::_convert_indices_from_coo_to_csr(
+      row_indices.view({-1}), block_size_0, false /* out_int32 */);
+  values = values.reshape({-1, values.size(-2), values.size(-1)});
+  not_zero_mask = not_zero_mask.reshape({-1});
+  // TODO: masked_select does not support some form of broadcasting, so we're
+  // using the mask to construct indices that are then passed into index_select.
+  // This isn't ideal.
+  values = values.index_select(
+      0,
+      at::native::arange(not_zero_mask.numel()).masked_select(not_zero_mask));
+
+  return at::native::_sparse_bsr_tensor_unsafe(
```

Comment:
I think the CI errors are explained by [_sparse_bsr_tensor_unsafe](https://github.com/pytorch/pytorch/blob/8b5f11c61eecd58214c631056a634f2eedc6455a/aten/src/ATen/native/sparse/SparseCsrTensor.cpp#L316-L326) taking const references to local variables as arguments that are going out of scope at function return. That also means things like the sizes are being deallocated and therefore we get the CI error we're seeing. We might need a version of _sparse_bsr_tensor_unsafe that accepts rvalues. Somehow the error does not persist, at least locally, when compiled with `DEBUG=1`

Reply:
No, that doesn't make any sense, because Tensors are reference counted and so shouldn't they be kept alive?